File: codeforces/pinely-round-2/e/e.cpp
#include <bits/stdc++.h>
using namespace std;

typedef long long ll;
typedef pair<int, int> pii;
typedef vector<int> vi;
typedef vector<vector<int> > vii;
#define trace(x) cerr << #x << " : " << x << endl
#define _ << " " << 
#define sz(x) ((int) (x).size())
#define X first
#define Y second

#define MAX_N 100005

// define variables here
int n, m, k;
vector <pii > edges;
vii adj, rev; 
vi hs, value;
vector <bool> color;
vector <ll> fvalue;


vector <int> range(int n) {
	vector <int> indices(n);
	for (int i = 0; i < n; ++i) indices[i] = i;
	return indices;
}

void read_input() {
	cin >> n >> m >> k;
	hs.clear();
	for (auto i: range(n)) {
		int h; cin >> h;
		hs.push_back(h);
	}
	edges.clear();
	for (auto i: range(m)) {
		int u, v; cin >> u >> v;
		edges.push_back({u - 1, v - 1});
	}
	// trace(hs[0] _ hs[1] _ sz(edges));
}

int dfs(int u) {
	color[u] = true;
	for (auto v: adj[u]) {
		if (!color[v]) dfs(v);
		int temp = 0;
		if (hs[v] > hs[u]) temp = 1;
		value[u] = max(value[u], value[v] + temp);
	}
	return value[u];
}

ll farthest_delayed(int u) {
	if (color[u]) return fvalue[u];
	color[u] = true;
	ll cur_v = 1LL * (value[u] + 1) * k + hs[u];
	ll ans = cur_v;
	for (auto v: rev[u]) {
		ll child_v = 1LL * value[v] * k + hs[v];
		if (cur_v > child_v) {
			ans = max(ans, farthest_delayed(v));
		}
	}
	return fvalue[u] = ans;
}

auto solve() {
	// main logic comes here
	adj = vii(n, vi());
	rev = vii(n, vi());
	// rev.clear(); rev.resize(n);
	for (auto [u, v]: edges) {
		adj[v].push_back(u);
		rev[u].push_back(v);
		// trace(u _ v);
	}
	value = vi(n, 0);
	color = vector <bool> (n, false);
	color.clear(); color.resize(n, false);
	for (auto u: range(n)) {
		if (!color[u]) {
			dfs(u);
		}
	}
	color.assign(n, false);
	fvalue = vector<ll> (n, 0);
	ll right = -1;
	vector <tuple <int, ll, int> > total;
	for (auto u: range(n)) {
		if (value[u] == 0) {
			total.push_back({hs[u], farthest_delayed(u), u});
		}
		right = max(right, 1LL * value[u] * k + hs[u]);
	}
	sort(total.begin(), total.end());
	ll ans = ll(1e16);
	for (auto[le, shifted, u]: total) {
		ans = min(ans, right - le);
		right = max(right, shifted);
	}
	return ans;
}

int main() {
	ios_base::sync_with_stdio(false); cin.tie(0);
	int t; cin >> t;
	while (t--) {
		read_input();
		auto ans = solve();
		cout << ans << endl;
	}
	return 0;
}

File: codeforces/pinely-round-2/e/e.py
from __future__ import annotations
from collections import Counter
from types import GeneratorType
# import sys
import os
import io
# from typing import Generic, TypeVar

# sys.setrecursionlimit(10**5)
input = io.BytesIO(os.read(0, os.fstat(0).st_size)).readline

## generated template for competetive programming
## problem E (5/9)
## codeforces, Pinely-Round-2
## author: @shamir0xe
## verdict: [AC/WA/TLE/MLE]


def bootstrap(f, stack=[]):
    def wrappedfunc(*args, **kwargs):
        if stack:
            return f(*args, **kwargs)
        else:
            to = f(*args, **kwargs)
            while True:
                if type(to) is GeneratorType:
                    stack.append(to)
                    to = next(to)
                else:
                    stack.pop()
                    if not stack:
                        break
                    to = stack[-1].send(to)
            return to

    return wrappedfunc


# define variables
n: int = 0
m: int
k: int
hs: list[int]
adj: list[list[int]]
rev: list[list[int]]
edges: list[tuple[int, int]]
color: list[bool]
value: list[int]
rights: list[int]
fvalue: list[int]


# define other functions here
@bootstrap
def dfs(u: int):
    color[u] = True
    # print(f"from {u + 1}")
    for v in adj[u]:
        # print(f"{u + 1}->{v + 1}")
        if not color[v]:
            yield dfs(v)
        temp = 0
        if hs[v] > hs[u]:
            temp = 1
        value[u] = max(value[u], value[v] + temp)
    # print(f"value[{u + 1}] = {value[u]}")
    yield


def get_max(u) -> int:
    color[u] = True
    ans = value[u] * k + hs[u]
    for v in [*adj[u], *rev[u]]:
        if not color[v]:
            ans = max(get_max(v), ans)
    return ans


@bootstrap
def farthest_delayed(u: int):
    if color[u]:
        yield fvalue[u]
    color[u] = True
    cur_v = (value[u] + 1) * k + hs[u]
    # print(f"{u + 1}, cur_days: {value[u]} cur_v: {cur_v}")
    ans = cur_v
    for v in rev[u]:
        child_v = value[v] * k + hs[v]
        if cur_v > child_v:
            yield farthest_delayed(v)
            ans = max(ans, fvalue[v])
    fvalue[u] = ans
    yield


class Mediator:
    def __init__(self) -> None:
        self.outputs: list[str] = []

    def read_input(self) -> Mediator:
        global n, m, k, hs, edges

        n, m, k = tuple(map(int, input().rstrip().split()))
        hs = list(map(int, input().rstrip().split()))
        edges = []
        for _ in range(m):
            u, v = tuple(map(int, input().rstrip().split()))
            u, v = u - 1, v - 1
            edges += [(u, v)]
        return self

    def solve(self) -> Mediator:
        """
        main logic comes here
        """
        global n, m, k, hs, edges, color, adj, value, rev, fvalue
        adj = [[] for _ in range(n)]
        rev = [[] for _ in range(n)]
        for u, v in edges:
            adj[v] += [u]
            rev[u] += [v]
        value = [0] * n
        color = [False] * n
        total: list[tuple[int, int, int]] = []
        for u in range(n):
            if not color[u]:
                dfs(u)
        color = [False] * n
        fvalue = [0] * n
        # for u in range(n):
        #     print(f"{u + 1} farthest_delayed: {farthest_delayed(u)}")
        right = -1

        for u in range(n):
            # print("~", u + 1, value[u], hs[u])
            # right = max(right, value[u] * k + hs[u])
            if value[u] == 0:
                farthest_delayed(u)
                total += [(hs[u], fvalue[u], u)]
            right = max(right, value[u] * k + hs[u])
        total = sorted(total)
        ans = int(1e19)
        for le, shifted, u in total:
            # print(f"[{u + 1}] {le}, {shifted} -- {right}")
            ans = min(ans, right - le)
            right = max(right, shifted)

        self.ans = ans
        return self

    def print_output(self) -> None:
        print(self.ans)
        # self.outputs += [str(self.ans)]


def main(test_case: bool = False):
    med = Mediator()

    def main_part():
        med.read_input().solve().print_output()

    if test_case:
        t = int(input())
        for _ in range(t):
            main_part()
        # print("\n".join(med.outputs))
    else:
        main_part()


if __name__ == "__main__":
    main(True)
